### ISSUE TITLE
System.ArgumentNullException when Parameter is Interface

### ISSUE THREAD

**Issue by rcfausto** [HangfireIO/Hangfire#695]
Hi Sergey,
We upgraded from 1.4.2 to 1.6.6. It's working fine if parameter is class. It still ok in 1.4.7 but not anymore in 1.5.0 and onward.

The parameter is being persisted in Hangfire.Job.Arguments column.

Tried what's in here https://discuss.hangfire.io/t/argument-as-interface-types/1189/7 but still the same.

Looks like it can't be deserialized on filter.

```csharp
...
GlobalJobFilters.Filters.Add(new JobActivatorAttribute());
...
...
public class JobActivatorAttribute : JobFilterAttribute,IClientFilter, IServerFilter, IElectStateFilter, IApplyStateFilter
{
public void OnPerforming(PerformingContext filterContext)
{
//filterContext.Job.Arguments[0] is null
}
}
```
Thank you.

**Comment by aidmsu** [comment]
Hi, @rcfausto 

Could you provide more information about the problem, please? What's a signature of the method which had to be performed? What's a value of Hangfire.Job.Arguments column for this job? Do you have exception stacktrace?

I guess the thing is that an information about type of passing arguments is not stored.  Did you try to set TypeNameHandling serialization option to TypeNameHandling.All? 
```csharp
JobHelper.SetSerializerSettings(
    new JsonSerializerSettings { TypeNameHandling = TypeNameHandling.All });
```

I reproduced the bug with default serialization options. But I can't reproduce it after changing TypeNameHandling serialization setting to TypeNameHandling.All

**Comment by odinserj** [comment]
@rcfausto, `Job.Arguments` collection contains serialized values. How did you work with the argument? How did you read its value?

**Comment by rcfausto** [comment]
Hi @odinserj 
I read it like this. 
var param = JObject.Parse(filterContext.Job.Arguments[0]);
var passedParameter = param.ToObject<PassedParameter>();

Hi @aidmsu 
I'm certain I tried it before but it didn't work. However I tried it now, it can now be deserialized.
Probably I didn't able to place that in windows service (acts as job processor), only in web app at that time.

Looks good now, thank you guys!

**Comment by odinserj** [comment]
Could you also tell me what is the type of the method parameter, and what's the real type of an argument you are passing? Are you using interfaces/abstract classes here?

**Comment by rcfausto** [comment]
It's like this:
public interface ISampleJobEngine
    {
        void UpdateMethod(IJobEngineParameters jobEngineParameters, IEnumerable<int> ids);
    }

class SampleJobEngine: ISampleJobEngine
    {
        public void UpdateMethod(IJobEngineParameters jobEngineParameters, IEnumerable<int> ids)
        {
            var x = 0;
        }
    }

**Comment by odinserj** [comment]
Interface arguments should be used only with the `TypeNameHandling.All` option, otherwise an exception should be thrown. However there was a bug in Hangfire (fixed in PR #731 today) that caused returning `null` values for interface types instead of throwing an exception. It allowed you to use methods incorrectly. So that's not a regression, it's a bug utilization caused this issue. The fix will be available in upcoming 1.6.7.

@rcfausto, thanks a lot for reporting this, and for all of the information!
